tests/components/recorder/test_init.py

```diff
@@ -1134,3 +1135,46 @@ def test_entity_id_filter(hass_recorder):
             db_events = list(session.query(Events).filter_by(event_type="hello"))
             # Keep referring idx + 1, as no new events are being added
             assert len(db_events) == idx + 1, data
+
+
+async def test_database_lock_and_unlock(hass: HomeAssistant, tmp_path):
+    """Test writing events during lock getting written after unlocking."""
+    # Use file DB, in memory DB cannot do write locks.
+    config = {recorder.CONF_DB_URL: "sqlite:///" + str(tmp_path / "pytest.db")}
+    await async_init_recorder_component(hass, config)
+    await hass.async_block_till_done()
+
+    instance: Recorder = hass.data[DATA_INSTANCE]
+
+    await instance.lock_database()
+
+    event_type = "EVENT_TEST"
+    event_data = {"test_attr": 5, "test_attr_10": "nice"}
+    hass.bus.fire(event_type, event_data)
+    task = asyncio.create_task(async_wait_recording_done(hass, instance))
+
+    # Recording can't be finished while lock is held
+    with pytest.raises(asyncio.TimeoutError):
+        await asyncio.wait_for(asyncio.shield(task), timeout=5)
+
+    with session_scope(hass=hass) as session:
+        db_events = list(session.query(Events).filter_by(event_type=event_type))
+        assert len(db_events) == 0
+
+    assert instance.unlock_database()
+
+    await task
+    with session_scope(hass=hass) as session:
+        db_events = list(session.query(Events).filter_by(event_type=event_type))
+        assert len(db_events) == 1
+
+
+async def test_database_lock_timeout(hass):
+    """Test locking database timeout when recorder stopped."""
+    await async_init_recorder_component(hass)
+    hass.bus.async_fire(EVENT_HOMEASSISTANT_STOP)
+    await hass.async_block_till_done()
+
+    instance: Recorder = hass.data[DATA_INSTANCE]
+    with pytest.raises(TimeoutError):
+        await instance.lock_database(lock_timeout=1)
```

Comment:
We can probably do a shorter timeout in tests?

Reply:
My worry is that the test might end up to timeout for other reasons than the database being locked (e.g. not all events got processed within the time). In the end this kind of timeout testing are somewhat racy by nature to some degree.

A quick test on my somewhat old laptop showed that it normally takes about 10ms to lock the database. I'll set it to 100ms which should be fairly save.